episodes/3-identify-the-problem.Rmd

```diff
@@ -0,0 +1,301 @@
+
+:::::::: challenge 
+As a group, name some potential problems or undesired outcomes from this graph...
+::::::::
+
+:::::::: solution
+
+ - The graph looks sparse, and unevenly so -- many bins have no observations
+ - Suggests that some years had more observations and others fewer based on somewhat arbitrary measurements (i.e. what calendar year happened to fall on)
+ - Hard to compare trends across time, or even subsequent years...
+
+::::::::
+
+As we discussed in the challenge, there are some issues to visualizing our data this way. A solution here might be to tinker with the bin width in the histogram code, but let's step back a minute. Do we necessarily need to dive into the depths of tinkering with the plot? We can evalulate this problem not in terms of the plot having a problem, but with our data type having a problem. There's  an opportunity to encode the observation times outside of coarse, somewhat arbitrary year groupings with the real, interpretable date they were collected. Let's do that using the tidyverse's 'lubridate' package. The important details here are that we are creating a 'datetime'-type variable using the recorded years, months, and days, which are currently all encoded as numeric types.
+
+```{r}
+krats <- rodents %>% filter(genus == "Dipodomys") #kangaroo rat genus
+dim(krats)
+
+krats <- krats %>% mutate(date = lubridate::ymd(paste(year,month,day,sep='-')))
+
+ggplot(krats, aes(date, fill=plot_type)) + 
+geom_histogram() +
+facet_wrap(~species)
+
+```
+
+This looks much better, and is easier to see the trends over time as well. Note our x axis still shows bins with year labelings, but the continuous spread of our data over these bins shows that dates are treated more continuously and fall more continuously within histogram bins.
+
+:::::::: callout 
+One aspect we can see with this exercise above is that by setting up a reproducible example, we can isolate the problem with data rather than simply asking a proximal problem (i.e. 'how can i change my plot to look like so'). This allows helpers and you to directly improve your code, but also allows the community to help in identifying the problem. You don't always need to  understand what exact lines of code or function calls are going wrong in order to get help!
+::::::::
+
+
+
+
+In general, we need to isolate the specific areas of code causing the bug or problem. There is no general rule of thumb as to how large this needs to be, but in general, think about what we would want to include in a reprex. Any early lines which we know run correctly and as intended may not need to be included, and we should seek to isolate the problem area as much as we can to make it understandable to others. 
+
+Let's add to our list of steps for identifying the problem: 
+
+0. Identify the problem area -- add print statements immediately upstream or downstream of problem areas, step into functions, and see whether any intermediate output can be further isolated. 
```

Comment:
The audience probably won't know what a "print statement" is or what it means to "step into functions". Do you think it's within the scope of this episode to cover those techniques?